dali/kernels/common/pad.h

```diff
@@ -0,0 +1,149 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_COMMON_PAD_H_
+#define DALI_KERNELS_COMMON_PAD_H_
+
+#include <cuda_runtime.h>
+#include <cstdint>
+#include <vector>
+#include "dali/kernels/alloc.h"
+#include "dali/core/span.h"
+#include "dali/core/common.h"
+#include "dali/core/error_handling.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/tensor_shape.h"
+#include "dali/core/static_switch.h"
+#include "dali/kernels/slice/slice_kernel_utils.h"
+#include "dali/kernels/slice/slice_flip_normalize_permute_common.h"
+#include "dali/kernels/slice/slice_flip_normalize_permute_pad_cuda_impl.cuh"
+
+namespace dali {
+namespace kernels {
+
+template <typename Type, size_t Dims>
+class DLL_PUBLIC PadGPU {
+ private:
+  static constexpr size_t kBlockDim = 512;
+  static constexpr size_t kBlockSize = 64 * kBlockDim;
+  size_t block_count_ = 0;
+
+ public:
+  DLL_PUBLIC PadGPU() = default;
+
+  /**
+  * @brief Find the biggest dimension on a given `axis`.
+  * This dimension `biggest_dim_size` will be the new dimension
+  * on the axis for all the samples of the batch.
+  */
+  DLL_PUBLIC KernelRequirements Setup(KernelContext &context,
+                                      const InListGPU<Type, Dims> &in,
+                                      int axis) {
+    KernelRequirements req;
+    if (in.num_elements() == 0) {
+      req.output_shapes = {in.shape};
+    } else {
```

Comment:
I would add another case, when `axis < 0` and out_shape = max(in_shapes) in all axes.

Reply:
> out_shape = max(in_shapes) in all axes.

by that you mean the no-op case when all the tensors have already the same shape?

Reply:
Or full padding on all the axis?

Reply:
Done